File: 611-valid-triangle-number/valid-triangle-number.cpp
class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        int cnt = 0, n = nums.size();
        sort(nums.begin(),nums.end());
        for (int i=0; i<n-2; i++){
            if (nums[i] <= 0) continue;
            for(int j=i+1; j<n-1; j++){
                auto val = std::upper_bound(nums.begin(), nums.end(), nums[i]+nums[j]-1);
                int idx = std::distance(nums.begin(), val);
                cnt += idx-j-1;
            }
        }
        return cnt;
    }
};

File: 611-valid-triangle-number/valid-triangle-number.py
class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        cnt,n = 0, len(nums)
        for i in range(n-2):
            if nums[i] <= 0: continue
            for j in range(i+1,n-1):
                idx = bisect_right(nums, nums[i]+nums[j]-1)-1
                cnt += idx-j
        return cnt